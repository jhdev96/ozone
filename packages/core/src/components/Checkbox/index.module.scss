@import '../../scss/root';


@keyframes ripple {
	0%   { transform: translate(0, 0); }
	80%  { transform: translate(20%, 20%) scale(1.6); }
	100% { 
    transform: translate(20%, 20%) scale(1.6); 
    opacity: 0; 
    border-radius: 0;
  }
}

@keyframes fade {
	0%   { opacity: 1; }
  100% { opacity: 0; }
}

.root {
  width: fit-content
}

.checkboxWrapper {
  display: block;
  position: relative;
  margin: 0.50rem 0;
  max-width: 15px;
  max-height: 15px;
  font-size: 22px;
  cursor: pointer;
}

.input {
  position: absolute;
  height: 0;
  width: 0;
  opacity: 0;
  cursor: pointer;
}

.input:checked ~ .checkbox {
  background-color: var(--blue-core);
  border-color: var(--blue-core);
  transition: background-color 0.8s;
}

.checkbox {
  position: absolute;
  width: 15px;
  height: 15px;
  border: 1px solid var(--gray-3);
  border-radius: 3px;
}

.input:checked ~ .checkbox:after {
  display: block;
}

.checkbox:after {
  content: "";
  position: absolute;
  display: none;
}

.checkbox:after {
  top: 1px;
  left: 4.5px;
  width: 2px;
  height: 7px;
  border: solid white;
  border-width: 0 2.5px 2.5px 0;
  transform: rotate(40deg);
}

.checkboxRipple {
  background-color: var(--blue-core);
  width: 10px;
  height: 10px;
  border-radius: 50%;
  animation: ripple 0.3s ease-out forwards, 
             fade 0.3s ease-out forwards;
}

.text {
  vertical-align: middle;
  margin-left: 25px;

  // theme
  &.light { color: var(--black-core); }
  &.dark { color: var(--gray-1); }
}

.input:disabled ~ .checkbox {
  background-color: rgba(0, 0, 0, 0.043);
  border: rgba(0, 0, 0, 0.461);
}

.input:disabled ~ .checkbox:after {
  border-color: var(--gray-4);
}

.error { 
  border-color: var(--red-core);
  color: var(--red-core) !important; 

  .checkboxRipple {
    background-color: var(--red-core) !important;
  }

  .checkbox { border-color: var(--red-1); }
  .input:checked ~ .checkbox {
    background-color: var(--red-core) !important;
    border-color: var(--red-core) !important;
  }

  ~ .text {
    color: var(--red-core) !important;
  }
}