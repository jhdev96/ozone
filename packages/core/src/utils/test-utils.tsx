import React from "react";


interface TestProps {
  dataTestId?: string;
}

export function DummyTestComponent({
  dataTestId = "test"
}: TestProps): JSX.Element 
{
  return (
    <span data-testid={dataTestId}>
      Hello
    </span>
  )
}

export function mockRefCurrent(currentProperties?: object) 
{
  const ref = {current: currentProperties};
  
  jest.spyOn(React, 'useRef').mockImplementation(() => {
    return ref;
  });
}