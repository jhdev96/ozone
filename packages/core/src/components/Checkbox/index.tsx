import React, { useState } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


interface CheckboxProps {
  /**
   * Styles the checkbox text to match a light
   * or dark theme.
   */
  theme?: ThemeType;
  /** 
   * Indicates whether the box is checked. 
   */
  isChecked?: boolean;
  /** 
   * Sets the `required` HTML attribute if true. 
   */
  isRequired?: boolean;
  /** 
   * Sets the `disabled` HTML attribute if true. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates visusally that there has been an error. 
   */
  hasError?: boolean;
  /** 
   * Gives a name to the checkbox for tracking input. 
   */
  name?: string;
  /** 
   * Callback to be invoked when a change is detected. 
   */
  onChange?: (checked: boolean, name?: string) => void;
  /**
   * A selector used in testing environments
   */
  dataTestId?: string;
  /** 
   * Content to be rendered beside the box. 
   */ 
  children?: React.ReactNode;
}

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  function Checkbox({ 
    theme       = 'light',
    isChecked   = false, 
    isRequired  = true, 
    isDisabled  = false, 
    hasError    = false,
    onChange    = () => {},
    name,
    dataTestId,
    children}: CheckboxProps,
    ref: React.ForwardedRef<HTMLInputElement>): JSX.Element
  {
    const [checked, setChecked] = useState(isChecked);

    const handleOnChange = () => {
      setChecked(!checked);
      onChange(checked, name);
    }

    const transition = checked && (
      <div className={styles.checkboxRipple} />
    )

    return (
      <label 
        className={styles.root} 
        htmlFor={name} 
        data-testid={dataTestId}
      >
        <div 
          className={classNames(
            styles.checkboxWrapper, 
            hasError && styles.error
          )} 
          role="checkbox"
        >
          <input
            className={styles.input} 
            type="checkbox"
            name={name}
            checked={checked}
            aria-checked={checked}
            required={isRequired}
            disabled={isDisabled}
            onChange={handleOnChange}
            ref={ref}
          />
          <span className={styles.checkbox}>
            {transition}
          </span>
        </div>
        <span 
          className={classNames(
            styles.text, 
            styles[theme]
          )}
        >
          {children}
        </span>
      </label>
    );
  }
)

export default Checkbox;

Checkbox.displayName = 'Checkbox';