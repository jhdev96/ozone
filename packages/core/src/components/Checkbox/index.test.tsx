import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Checkbox from '.';


describe('<Checkbox />', () => {
  it('renders a checkbox with text', () => {
    render(
      <Checkbox dataTestId="test">text</Checkbox>
    );

    expect(screen.getByTestId('test'))
      .toHaveTextContent('text');
  });

  it('sets `checked` attribute to true when clicked', () => {
    
    act(() => {
      render(
        <Checkbox dataTestId="test">text</Checkbox>
      );
    });

    const checkbox = screen.getByTestId('test');
  
    fireEvent.click(checkbox);
    expect(checkbox.querySelector('input')!.checked)
      .toBe(true);
  });

  it('renders a transition when the input is clicked', () => {
    act(() => {
      render(
        <Checkbox dataTestId="test">text</Checkbox>
      );
    });

    const checkbox = screen.getByTestId('test');

    fireEvent.click(checkbox);
    expect(checkbox.querySelector('.checkboxRipple'))
      .toBeInTheDocument;
  });

  it('sets the `required` attribute to true', () => {
    render(
      <Checkbox dataTestId="test" isRequired>text</Checkbox>
    );

    const checkbox = screen.getByTestId('test'); 

    expect(checkbox.querySelector('input')!.required)
      .toBe(true);
  });

  it('renders a disabled checkbox', () => {
    render(
      <Checkbox dataTestId="test" isDisabled>text</Checkbox>
    );

    const checkbox = screen.getByTestId('test');

    expect(checkbox.querySelector('input')!.disabled)
      .toBe(true);
  });

  it('renders a checkbox with an error', () => {
    render(
      <Checkbox 
        dataTestId="test" 
        hasError
      >
        text
      </Checkbox>
    );

    const checkbox = screen.getByTestId('test');

    expect(checkbox).toMatchSnapshot();
  });

  it('calls `onChange` when the checkbox is clicked', () => {
    const onChange = jest.fn();

    act(() => {
      render(
        <Checkbox 
          onChange={onChange} 
          dataTestId="test"
        >
          text
        </Checkbox>
      );
    });

    const checkbox = screen.getByTestId('test');

    fireEvent.click(checkbox);
    expect(onChange).toBeCalled;
  });
});