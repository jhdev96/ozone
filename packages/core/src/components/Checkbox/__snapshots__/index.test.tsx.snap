// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders a checkbox with an error 1`] = `
<label
  class="root"
  data-testid="test"
>
  <div
    class="checkboxWrapper error"
    role="checkbox"
  >
    <input
      aria-checked="false"
      class="input"
      required=""
      type="checkbox"
    />
    <span
      class="checkbox"
    />
  </div>
  <span
    class="text light"
  >
    text
  </span>
</label>
`;
